src/hyperplane/aws_s3_utils.py

```diff
@@ -0,0 +1,200 @@
+from hyperplane import get_s3_credentials, report
+
+
+def _s3_client(access_key_id=None, access_key_secret=None):
+    if not access_key_id and not access_key_secret:
+        return boto3.client('s3')
+    session = _s3_session_with_creds(access_key_id, access_key_secret)
+    return session.client('s3')
+
+
+def _s3_resource(access_key_id, access_key_secret):
+    if not access_key_id and not access_key_secret:
+        return boto3.resource('s3')
+    session = _s3_session_with_creds(access_key_id, access_key_secret)
+    return session.resource('s3')
+
+
+def _s3_session_with_creds(access_key_id, access_key_secret):
+    return boto3.Session(
+        aws_access_key_id=access_key_id,
+        aws_secret_access_key=access_key_secret,
+    )
+
+
+def _connect_to_s3_bucket(bucket_name, user_s3_creds, access_key_secret=None, access_key_id=None,
+                          s3_client_or_resource=_s3_client):
+    print(f"Connecting s3 bucket: {bucket_name}")
+    if access_key_id and access_key_secret:
+        print("Connecting to s3 bucket with given access key id and secret access key")
+        return s3_client_or_resource(access_key_id, access_key_secret)
+
+    user_bucket_name = user_s3_creds.get("bucket_url")
+    if bucket_name == user_bucket_name:
+        if not access_key_secret:
+            access_key_secret = user_s3_creds.get("secret_access_key")
+        if not access_key_id:
+            access_key_id = user_s3_creds.get("access_key_id")
+
+    return s3_client_or_resource(access_key_id, access_key_secret)
+
+
+def _try_get_s3_creds_from_env():
+    access_key_id = os.environ.get('HYPERPLANE_S3_ACCESS_KEY_ID')
+    access_key_secret = os.environ.get('HYPERPLANE_S3_ACCESS_KEY_SECRET')
+
+    return access_key_id, access_key_secret
+
+
+def download_from_s3(source_path: str, target_path: str = None, regex_filter: str = None):
+    """
+    Args:
+        source_path: for directories put '/' in the end
+        target_path: for dir: Should end with '/'. files from source_path dir will be uploaded under target_path.
+                     for file: where the file will be saved (including name).
+        regex_filter: filter by regex
+    """
+    user_s3_creds = get_s3_credentials()
+
+    bucket_name = user_s3_creds.get("bucket_url")
+    if not bucket_name:
+        report("User didn't specify bucket name and have no bucket configurations in our system")
+        return None
+    s3_resource = _connect_to_s3_bucket(bucket_name=bucket_name, access_key_id=user_s3_creds.get("access_key_id"),
+                                        access_key_secret=user_s3_creds.get("secret_access_key"),
+                                        s3_client_or_resource=_s3_resource,
+                                        user_s3_creds=user_s3_creds)
+
+    bucket = s3_resource.Bucket(bucket_name)
+    objects_to_download = list(bucket.objects.filter(Prefix=source_path))
+    regex_compile = re.compile(regex_filter) if regex_filter else None
```

Comment:
compile this outside of this function, it makes no sense to compile every time for every file (there could be 1000's of them)

Reply:
it is outside from the for loop so...